src/shapes/group.class.js

```diff
@@ -146,25 +136,68 @@
      * @chainable
      */
     addWithUpdate: function(object) {
+      return this.insertWithUpdate(object);
+    },
+
+    /**
+     * Inserts an object into collection at specified index; Then recalculates group's dimension, position.
+     * @param {Object} object Object to insert
+     * @param {Number} index Index to insert object at
+     * @param {Boolean} nonSplicing When `true`, no splicing (shifting) of objects occurs
+     * @return {fabric.Group} thisArg
+     * @chainable
+     */
+    insertWithUpdate: function(object, index, nonSplicing) {
+      if (this.group) {
```

Comment:
All this parent group checking and index maintaining is a workaround for the issue I described here: https://github.com/kangax/fabric.js/pull/2997#issuecomment-221310801

This solves that problem but probably in an inelegant way. It can be removed when a better fix is found.